src/libraries/System.Private.CoreLib/src/System/UInt128.cs

```diff
@@ -0,0 +1,1835 @@
+        public static explicit operator nuint(UInt128 value) => (nuint)value._lower;
+
+        /// <summary>Explicitly converts a 128-bit unsigned integer to a <see cref="UIntPtr" /> value, throwing an overflow exception for any values that fall outside the representable range.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to a <see cref="UIntPtr" />.</returns>
+        /// <exception cref="OverflowException"><paramref name="value" /> is not representable by <see cref="UInt128" />.</exception>
+        [CLSCompliant(false)]
+        public static explicit operator checked nuint(UInt128 value)
+        {
+            if (value._upper != 0)
+            {
+                ThrowHelper.ThrowOverflowException();
+            }
+            return checked((nuint)value._lower);
+        }
+
+        //
+        // Explicit Conversions To UInt128
+        //
+
+        /// <summary>Explicitly converts a <see cref="decimal" /> value to a 128-bit signed integer.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to a 128-bit signed integer.</returns>
+        public static explicit operator UInt128(decimal value)
+        {
+            value = decimal.Truncate(value);
+
+            if (value < 0.0m)
+            {
+                ThrowHelper.ThrowOverflowException();
+            }
+            return new UInt128(value.High, value.Low64);
+        }
+
+        /// <summary>Explicitly converts a <see cref="double" /> value to a 128-bit signed integer.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to a 128-bit signed integer.</returns>
+        public static explicit operator UInt128(double value)
+        {
+            const double TwoPow128 = 340282366920938463463374607431768211456.0;
+
+            if (double.IsNegative(value) || double.IsNaN(value))
+            {
+                return MinValue;
+            }
+            else if (value >= TwoPow128)
+            {
+                return MaxValue;
+            }
+
+            return ToUInt128(value);
+        }
+
+        /// <summary>Explicitly converts a <see cref="double" /> value to a 128-bit signed integer, throwing an overflow exception for any values that fall outside the representable range.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to a 128-bit signed integer.</returns>
+        /// <exception cref="OverflowException"><paramref name="value" /> is not representable by <see cref="UInt128" />.</exception>
+        public static explicit operator checked UInt128(double value)
+        {
+            const double TwoPow128 = 340282366920938463463374607431768211456.0;
+
+            if ((value < 0.0) || double.IsNaN(value) || (value >= TwoPow128))
```

Comment:
This is because of `-0`. Basically for the purposes of `floating-point` to `unsigned integer` conversions, `-0` is treated as `+0` and not as an invalid negative number.

I'll add a comment.